File: Domains/Base/Workspace/Typescript/Domain/src/workspace/Workspace.ts
import { ObjectType } from "@allors/base-meta";
import { MetaPopulation } from "@allors/base-meta";

import { PullResponse } from "./../database/pull/PullResponse";
import { SyncRequest } from "./../database/sync/SyncRequest";
import { SyncResponse } from "./../database/sync/SyncResponse";

import { SessionObject } from "./SessionObject";
import { WorkspaceObject } from "./WorkspaceObject";

export interface IWorkspace {
    metaPopulation: MetaPopulation;
    constructorByName: { [name: string]: typeof SessionObject };
    diff(data: PullResponse): SyncRequest;
    sync(data: SyncResponse): void;
    get(id: string): WorkspaceObject;
}

export class Workspace implements IWorkspace {

    private workspaceObjectById: { [id: string]: WorkspaceObject; } = {};

    constructor(public metaPopulation: MetaPopulation, public constructorByName: { [name: string]: typeof SessionObject }) {
    }

    public diff(response: PullResponse): SyncRequest {
        const requireLoadIds = new SyncRequest();

        if (response.objects) {
            const userSecurityHash = response.userSecurityHash;

            const requireLoadIdsWithVersion = response.objects.filter((idAndVersion) => {
                const [id, version] = idAndVersion;
                const workspaceObject = this.workspaceObjectById[id];

                return (workspaceObject === undefined) ||
                    (workspaceObject === null) ||
                    (workspaceObject.version !== version) ||
                    (workspaceObject.userSecurityHash !== userSecurityHash);
            });

            requireLoadIds.objects = requireLoadIdsWithVersion.map((idWithVersion) => {
                return idWithVersion[0];
            });
        }

        return requireLoadIds;
    }

    public sync(syncResponse: SyncResponse): void {
        if (syncResponse.objects) {
            Object
                .keys(syncResponse.objects)
                .forEach((v) => {
                    const objectData = syncResponse.objects[v];
                    const workspaceObject = new WorkspaceObject(this, syncResponse, objectData);
                    this.workspaceObjectById[workspaceObject.id] = workspaceObject;
                });
        }
    }

    public get(id: string): WorkspaceObject {
        const workspaceObject = this.workspaceObjectById[id];
        if (workspaceObject === undefined) {
            throw new Error(`Object with id ${id} is not present.`);
        }

        return workspaceObject;
    }
}

File: Domains/Base/Workspace/CSharp/Domain/Base/Workspace/Workspace.cs
namespace Allors.Workspace
{
    using System;
    using System.Collections.Generic;
    using System.Linq;

    using Allors.Protocol.Remote.Pull;
    using Allors.Protocol.Remote.Sync;

    public class Workspace
    {
        private readonly Dictionary<long, WorkspaceObject> workspaceObjectById = new Dictionary<long, WorkspaceObject>();
        
        public Workspace(ObjectFactory objectFactory)
        {
            this.ObjectFactory = objectFactory;
        }

        public ObjectFactory ObjectFactory { get; }

        public SyncRequest Diff(PullResponse response)
        {
            var userSecurityHash = response.UserSecurityHash;

            var requireLoadIds = new SyncRequest
                                     {
                                         Objects = response.Objects.Where(v =>
                                                 {
                                                     var id = long.Parse(v[0]);
                                                     var version = long.Parse(v[1]);
                                                     this.workspaceObjectById.TryGetValue(id, out var workspaceObject);
                                                     return workspaceObject == null || !workspaceObject.Version.Equals(version) || !workspaceObject.UserSecurityHash.Equals(userSecurityHash);
                                                 }).Select(v => v[0]).ToArray()
                                     };

            return requireLoadIds;
        }

        public void Sync(SyncResponse syncResponse)
        {
            foreach (var objectData in syncResponse.Objects)
            {
                var workspaceObject = new WorkspaceObject(this, syncResponse, objectData);
                this.workspaceObjectById[workspaceObject.Id] = workspaceObject;
            }
        }

        public WorkspaceObject Get(long id)
        {
            var workspaceObject = this.workspaceObjectById[id];
            if (workspaceObject == null)
            {
                throw new Exception($"Object with id {id} is not present.");
            }

            return workspaceObject;
        }
    }
}